web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte

```diff
@@ -0,0 +1,147 @@
+  import {
+    Divider,
+    Menu,
+    MenuItem,
+  } from "@rilldata/web-common/components/menu";
+  import Tooltip from "@rilldata/web-common/components/tooltip/Tooltip.svelte";
+  import TooltipContent from "@rilldata/web-common/components/tooltip/TooltipContent.svelte";
+  import type { MetricsViewDimension } from "@rilldata/web-common/runtime-client";
+  import { NO_COMPARISON_LABEL } from "@rilldata/web-common/lib/time/config";
+  import SelectorButton from "./SelectorButton.svelte";
+  import Compare from "@rilldata/web-common/components/icons/Compare.svelte";
+  import ClockCircle from "@rilldata/web-common/components/icons/ClockCircle.svelte";
+  import { Search } from "@rilldata/web-common/components/search";
+
+  const dispatch = createEventDispatcher();
+
+  export let showTimeComparison = true;
+  export let selectedDimension;
+  export let dimensions: MetricsViewDimension[];
+
+  const TIME = "Time";
+
+  let searchText = "";
+
+  function getLabelForDimension(dimension: string) {
+    return dimensions.find((d) => d.name === dimension)?.label;
+  }
+
+  /** compile the comparison options */
+  let options = dimensions.map((d) => ({
+    name: d.name,
+    label: d.label,
+  }));
+
+  $: menuOptions = matchSorter(options, searchText, { keys: ["label"] });
+
+  $: label = selectedDimension
+    ? getLabelForDimension(selectedDimension)
+    : showTimeComparison
+    ? TIME
+    : NO_COMPARISON_LABEL;
+
+  $: intermediateSelection = selectedDimension
+    ? selectedDimension
+    : showTimeComparison
+    ? TIME
+    : NO_COMPARISON_LABEL;
+</script>
+
+<WithTogglableFloatingElement
+  distance={8}
+  alignment="start"
+  let:toggleFloatingElement
+  let:active
+>
+  <Tooltip distance={8} suppress={active}>
+    <SelectorButton
+      {active}
+      on:click={() => {
+        toggleFloatingElement();
+      }}
+    >
+      <div class="flex items-center gap-x-3">
+        <span class="ui-copy-icon"><Compare size="16px" /></span>
+
+        <span style:transform="translateY(-1px)" class="font-normal">
+          {showTimeComparison || selectedDimension ? "Comparing by" : ""}
+          <span class="font-bold">{label}</span>
+        </span>
+      </div>
+    </SelectorButton>
+    <TooltipContent slot="tooltip-content" maxWidth="220px">
+      Select a comparison for the dashboard
+    </TooltipContent>
+  </Tooltip>
+  <Menu
+    minWidth="280px"
+    slot="floating-element"
+    on:escape={toggleFloatingElement}
+    on:click-outside={toggleFloatingElement}
+    label="Comparison selector"
+  >
+    <Search placeholder="Search Dimension" bind:value={searchText} />
+    <MenuItem
+      focusOnMount={false}
+      selected={!(showTimeComparison || selectedDimension)}
+      on:before-select={() => {
+        intermediateSelection = NO_COMPARISON_LABEL;
+      }}
+      on:select={() => {
+        dispatch("disable-all-comparison");
+        toggleFloatingElement();
+      }}
```

Comment:
This `on:before-select` and `intermediateSelection` looks over-complicated, but looks like you're just using the `MenuItem` component as designed?

Reply:
Yes, it's pretty much the same as other menu components. IIRC this was done to change the state in UI without waiting for the state loop to come back in and change it. For slower dashboards, we wanted to show the change in selection as soon as possible without waiting for all the API calls to resolve.